future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+     */
+    function calculatePayrollRunway() public view returns(uint256 payrollRunway) {
+        if (yearlyTotalPayroll == 0)
+            return MAX_UINT;
+
+        uint256 balance = 0;
+        // for each token, get finance balance and convert it to Denomination
+        for (uint i = 0; i < allowedTokensArray.length; i ++) {
+            var token = allowedTokensArray[i];
+            var tokenContract = ERC20(token);
+            if (exchangeRates[token] == 0)
+                continue;
+            balance = balance.add(tokenContract.balanceOf(finance).div(exchangeRates[token]));
+        }
+        payrollRunway = balance.mul(365) / yearlyTotalPayroll;
+    }
+
+    /**
+     * @dev Set token distribution for payments to an employee (the caller)
+     * @notice Set token distribution for payments to an employee (the caller).
+     *         Only callable once every 6 months
+     * @param tokens Array with the tokens to receive, they must belong to allowed tokens for employee
+     * @param distribution Array (correlated to tokens) with the proportions (integers over 100)
+     */
+    function determineAllocation(address[] tokens, uint256[] distribution) public {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+        // check that enough time has gone by
+        require(getTimestamp() > employee.lastAllocation && getTimestamp() - employee.lastAllocation > 15768000); // half a year in seconds
+
+        // check arrays match
+        require(tokens.length == distribution.length);
+
+        // check distribution is right
+        uint256 sum = 0;
+        uint256 i;
+        for (i = 0; i < distribution.length; i++) {
+            // check token is allowed
+            require(allowedTokens[tokens[i]]);
+            // set distribution
+            employee.allocation[tokens[i]] = distribution[i];
+            sum = sum.add(distribution[i]);
+        }
+        require(sum == 100);
+
+        employee.lastAllocation = getTimestamp();
+    }
+
+    /**
+     * @dev Get payment proportion for a token and an employee (the caller)
+     * @notice Get payment proportion for a token and an employee (the caller)
+     * @param token The token address
+     */
+    function getAllocation(address token) public view returns(uint256 allocation) {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+
+        allocation = employee.allocation[token];
+    }
+
+    /**
+     * @dev payday To withdraw monthly payment by employee (the caller)
+     * @notice Only callable once a month. Assumed token has these standard checks implemented:
+     *         https://theethereum.wiki/w/index.php/ERC20_Token_Standard#How_Does_A_Token_Contract_Work.3F
+     */
+    function payday() public {
```

Comment:
This function is buried too deep, we try to follow this as much as possible: http://solidity.readthedocs.io/en/develop/style-guide.html#order-of-functions

Reply:
Also should be external